gridded_etl_tools/utils/publish.py

```diff
@@ -702,6 +710,92 @@ def post_parse_quality_check(self, checks: int = 100, threshold: float = 10e-5):
                 f"Checking dataset took {datetime.timedelta(seconds=elapsed)}"
             )
 
+    def filter_search_space(self, prod_ds: xr.Dataset) -> tuple[pathlib.Path]:
+        """
+        Filter down all input files to only files that are within the update date range
+        on the production dataset.
+
+        NOTE this implicitly relies on input files being sorted by the time dimension,
+        so that the determined bounds encapsulate the entire range of valid data.
+
+        Parameters
+        ----------
+        prod_ds : xr.Dataset
+            The production dataset, filtered down to the most recent update
+
+        Returns
+        -------
+        tuple[pathlib.Path]
+            A list of valid input files in pathlib.Path format
+        """
+        possible_files = list(self.input_files())
+        production_date_range = self.strings_to_date_range(prod_ds.attrs["update_date_range"])
+
+        # Determine an upper bound using a binary search
+        left, right = 0, len(possible_files) - 1
+        upper_bound_index = -1
+
+        while left <= right:
+            mid = (left + right) // 2
+            if self.compare_dates(production_date_range, possible_files[mid]):
+                upper_bound_index = mid
+                right = mid - 1
+            else:
+                left = mid + 1
+
+        # Only return file paths within these bounds
+        possible_files = possible_files[upper_bound_index:]
+        # Apply any additional, custom filtering on the way out
+        return self.custom_file_filter(possible_files)
+
+    def compare_dates(
+        self, production_date_range: tuple[datetime.datetime, datetime.datetime], orig_file_path: pathlib.Path
+    ) -> bool:
+        """
+        Assess whether the date in the selected original dataset falls witihn the production
+
+        Parameters
+        ----------
+        production_date_range : tuple[datetime.datetime, datetime.datetime]
+            The date range for the production dataset
+        orig_file_path : pathlib.Path
+            The path to the original file, on disk or remotely
+
+        Returns
+        -------
+        bool
+            _description_
+        """
+        orig_ds = self.raw_file_to_dataset(orig_file_path)
+
+        if (
+            production_date_range[0]
+            <= self.numpydate_to_py(orig_ds[self.time_dim].values[0])
+            <= production_date_range[1]
+        ):
+            return True
+        else:
+            return False
+
+    def custom_file_filter(self, original_files: tuple[str]) -> tuple[str]:
```

Comment:
I still think we don't need custom_file_filter. The convention we're already using where we override the existing function is better design-wise.

```py
class Publish:

    def filter_search_space(self):
        return ["file0.nc4", "file1.nc4", "file2.nc4"]

class MyETL:

    def filter_search_space(self):
        possible_files = super().filter_search_space()
        return possible_files[1:] // the customization is to always remove the first file
```

Reply:
OK this makes sense, I'm implementing it so